### PR TITLE
OpenSSL engine

### PR DESCRIPTION
**Pull request by ekoby**

### CODE REVIEW

File: src/tls_link.c [hunk 1]
```diff
@@ -143,6 +148,13 @@ static void tls_read_cb(uv_link_t *l, ssize_t nread, const uv_buf_t *b) {
             inlen = 0;
         } while (rc == TLS_MORE_AVAILABLE && out_bytes > 0);
 
+        if (rc == TLS_HAS_WRITE) {
+            uv_buf_t buf;
+            buf.base = malloc(32 * 1024);
+            int tls_rc = tls->engine->api->write(tls->engine->engine, NULL, 0, buf.base, &buf.len, 32 * 1024);
```
Comment (dovholuknf) [inline src/tls_link.c:154]: nit. probably should be a constant. someone might miss that 32 * 1024 is there twice

File: src/tls_engine.c [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+static tls_context_factory factory = NULL;
+#endif
+
+void set_default_tls_impl(tls_context_factory f) {
+    factory = f;
+}
+
+tls_context *default_tls_context(const char *ca, size_t ca_len) {
+    if (factory == NULL) {
+        UM_LOG(ERR, "FATAL error no default TLS engine is set");
```
Comment (dovholuknf) [inline src/tls_engine.c:36]: would it be helpful to actually point someone to 'how' to set the engine  here?

File: .github/workflows/cmake.yml [hunk 3]
```diff
@@ -24,7 +29,7 @@ jobs:
     - name: make build dir
       run: cmake -E make_directory ${{runner.workspace}}/build
     - name: configure cmake
-      run: cmake -S ${{ github.workspace }} -B ${{runner.workspace}}/build
+      run: cmake -S ${{ github.workspace }} -B ${{runner.workspace}}/build -DUSE_OPENSSL=${{ matrix.openssl }}
```
Comment (dovholuknf) [inline .github/workflows/cmake.yml:32]: do you think it'd be useful to test it with both options? Seems like we should test it both ways?

File: CMakeLists.txt [hunk 4]
```diff
@@ -7,15 +7,35 @@ PROJECT("uv-mbed"
...
 option(ENABLE_UM_TESTS "Build tests." ON)
 
+option(USE_OPENSSL "use OpenSSL engine" OFF)
+
+if(NOT USE_OPENSSL)
+    message("using mbedTLS libraries")
+    set(USE_MBEDTLS true)
+    set(tlsImpl mbedtls)
+else()
+    set(tlsImpl openssl)
```
Comment (dovholuknf) [inline CMakeLists.txt:19]: no "using openssl" message?

File: deps/mbedtls.cmake [hunk 5]
```diff
@@ -0,0 +1,10 @@
+include(FetchContent)
+FetchContent_Declare(mbedtls
+        GIT_REPOSITORY https://github.com/netfoundry/mbedtls.git
+        GIT_TAG verify-ip-sans-properly
```
Comment (dovholuknf) [inline deps/mbedtls.cmake:4]: that's really the tag or is that a holdover?

File: .github/workflows/cmake.yml (hunk 3 above)
Comment (ekoby) [reply .github/workflows/cmake.yml:32]: it does use both options. see job.matrix

File: deps/mbedtls.cmake (hunk 5 above)
Comment (ekoby) [reply deps/mbedtls.cmake:4]: branch

**Pull request merged**